base/stats.go

```diff
@@ -16,13 +52,39 @@ const (
 //to find valid property names
 //
 func init() {
-	StatsExpvars.Add("changesFeeds_total", 0)
-	StatsExpvars.Add("changesFeeds_active", 0)
-	StatsExpvars.Add("requests_total", 0)
-	StatsExpvars.Add("requests_active", 0)
-	StatsExpvars.Add("revisionCache_hits", 0)
-	StatsExpvars.Add("revisionCache_misses", 0)
+	for _, stat := range globalStatNames {
+		StatsExpvars.Add(stat, 0)
+	}
 	TimingExpvars = NewSequenceTimingExpvar(KTimingExpvarFrequency, KTimingExpvarVbNo, "st")
 	StatsExpvars.Set("sequenceTiming", TimingExpvars)
+	InitDbStats("_server")
+
+	ticker := time.NewTicker(statLogInterval)
+	go func () {
+		for {
+			<-ticker.C
+			Statsf(KeyAll, StatsExpvars.String())
+		}
+	}()
+}
+
+func InitDbStats(name string) {
+	if StatsExpvars.Get(name) == nil {
+		dbStats := expvar.NewMap(name)
+		for _, stat := range perDbStatNames {
+			dbStats.Add(stat, 0)
+		}
+		StatsExpvars.Set(name, dbStats)
+	}
+}
 
+func UpdateDbStat(dbName, statName string, delta int64) {
+	if dbName != "" {
+		dbStats := StatsExpvars.Get(dbName)
+		if s, ok := dbStats.(*expvar.Map); ok {
+			s.Add(statName, delta)
+		}
+	}
+	// Update the global stat. Do this even if it doesn't correspond to a db for backwards compatibility
+	StatsExpvars.Add(statName, delta)
```

Comment:
For performance, a reference to dbStats should be added to the databaseContext, instead of looking it up by name from StatsExpvars every time this is called.  UpdateDbStat would then be databaseContext.UpdateStat.  The backwards compatibility call would remain the same.

Reply:
Sounds good. I didn't want to start meddling potentially unrelated things at first though ;)